architecture/authentication.adoc

```diff
@@ -0,0 +1,268 @@
+
+==== Identify users [[config-request-handlers]]
+
+To configure how the OAuth server recognizes users from incoming requests to `/oauth/authorize`, set the `ORIGIN_OAUTH_REQUEST_HANDLERS` environment variable. Valid values are:
+
+* `requestheader`
+** Identify the user from a configurable request header (e.g. `X-Remote-User`)
+** Typically used with an authenticating proxy which sets the request header
+** Uses the specified link:#config-requestheader[request header configuration]
+* `bearer`
+** Identify the user from an "Authorization: Bearer" header, if present
+** Uses the configured link:#config-bearer[bearer token validation]
+* `basicauth`
+** Identify the user from an "Authorization: Basic" header, if present
+** Uses the configured link:#config-password[username and password validation]
+* `session`
+** Identify the user from a cookie-based session
+** Uses the specified link:#config-session[session configuration]
+
+Multiple handlers can be specified using a comma-separated list. Each handler is given a chance to identify the user from the request. The first handler to identify the user wins.
+
+`ORIGIN_OAUTH_REQUEST_HANDLERS` defaults to `session,basicauth`
+
+==== Handle unauthenticated requests [[config-auth-handler]]
+
+To configure how the OAuth server responds to unauthenticated requests to `/oauth/authorize`, set the `ORIGIN_OAUTH_HANDLER` environment variable. Valid values are:
+
+* `login`
+** Redirect to an OpenShift-provided login page, and validate the login using the configured link:#config-password[username and password validation]
+** For OAuth clients that request it, issue a `WWW-Authenticate: Basic` challenge rather than redirecting
+** Requires `session` and `basicauth` to be in the `ORIGIN_OAUTH_REQUEST_HANDLERS` list
+* `github`
+** Allows logging in using GitHub credentials
+** Uses the specified link:#config-github[Github OAuth client configuration]
+** Requires `session` to be in the `ORIGIN_OAUTH_REQUEST_HANDLERS` list
+** Using an external identity provider via OAuth prevents non-browser clients from obtaining API tokens directly
+* `google`
+** Allows logging in using Google credentials
+** Uses the specified link:#config-google[Google OAuth client configuration]
+** Requires `session` to be in the `ORIGIN_OAUTH_REQUEST_HANDLERS` list
+** Using an external identity provider via OAuth prevents non-browser clients from obtaining API tokens directly
+* `deny`
+** Reject unauthenticated requests
+
+`ORIGIN_OAUTH_HANDLER` defaults to `login`
+
+==== Bearer token validation [[config-bearer]]
+
+This configuration is used by the link:#config-request-handlers[`bearer`] request handler.
+
+To configure how the OAuth server validates bearer tokens, set the `ORIGIN_OAUTH_TOKEN_STORE` environment variable. Valid values are:
+
+* `etcd`
+** Bearer tokens must be valid OAuth access tokens (which are stored in etcd)
+** This allows a user with an API token to generate another API token
+* `file`
```

Comment:
I don't think we ever want someone to use anything other than etcd.  Can you think of a solid use-case for exposing this?